### PR TITLE
Added Monitoring Capabilities (Observability-as-code)

### PR DESCRIPTION
**Pull request by sudivate**
## Summary of the Pull Request
Added below capabilities that can be optionally disabled.
- Azure Dashboard, alerts and diagnostic settings for DataFactory and Synapse, SQL DB's(MSSQL,MYSQL,MariaDB,PostgreSQL, Cosmos DB
- Log analytics Workspace, Azure Data Factory Analytics Solution, Azure SQL Analytics Solution
- Updated Portal Experience  
- Updated Doc 

## References
No reference

## PR Checklist
* [x ] Code of Conduct signed. If not, go to [Code of Conduct](https://github.com/Azure/data-management-zone/blob/main/CODE_OF_CONDUCT.md).

## Validation Steps Performed
Deployed manually

### CODE REVIEW

File: docs/reference/portal.dataProduct.json [hunk 1]
```diff
@@ -595,6 +595,54 @@
...
+                                    "type": "Microsoft.Common.OptionsGroup",
+                                    "visible": true,
+                                    "toolTip": "Deploy monitoring components like Log Analytics, Azure dashboards, alerts and diagnostic settings.",
+                                    "defaultValue": "No",
+                                    "constraints": {
+                                        "allowedValues": [
+                                            {
+                                                "label": "Yes",
+                                                "value": "yes"
+                                            },
```
Comment (marvinbuss) [inline docs/reference/portal.dataProduct.json:629]:
```suggestion
                                    "defaultValue": "Yes (recommended)",
                                    "constraints": {
                                        "allowedValues": [
                                            {
                                                "label": "Yes (recommended)",
                                                "value": "yes"
                                            },
```

Comment (marvinbuss) [reply docs/reference/portal.dataProduct.json:629]: I guess we should enable this by default?

File: docs/reference/portal.dataProduct.json [hunk 2]
```diff
@@ -595,6 +595,54 @@
...
+                                    }
+                                },
+                                {
+                                    "name": "dataProductTeamEmail",
+                                    "label": "Alert Receivers Email ID",
+                                    "type": "Microsoft.Common.TextBox",
+                                    "visible": "[equals(steps('generalSettings').monitoringSettings.enableMonitoring, 'yes')]",
+                                    "defaultValue": "",
+                                    "toolTip": "Specify Email ID to receive monitoring alerts."
+                                }
```
Comment (marvinbuss) [inline docs/reference/portal.dataProduct.json:644]: Can we add some regex validation here to make sure that an email was entered and also specify whether this is required or not?

File: infra/main.bicep [hunk 3]
```diff
@@ -97,10 +103,17 @@ var keyVault001Name = '${name}-vault001'
 var synapse001Name = '${name}-synapse001'
 var datafactory001Name = '${name}-datafactory001'
 var cosmosdb001Name = '${name}-cosmos001'
+var Database001Name = 'Database001'
```
Comment (marvinbuss) [inline infra/main.bicep:106]:
```suggestion
var database001Name = 'Database001'
```

File: infra/main.bicep [hunk 4]
```diff
@@ -185,6 +198,7 @@ module sql001 'modules/services/sql.bicep' = if (sqlFlavour == 'sql') {
     administratorUsername: administratorUsername
     administratorPassword: administratorPassword
     privateDnsZoneIdSqlServer: privateDnsZoneIdSqlServer
+    Database001Name: Database001Name
```
Comment (marvinbuss) [inline infra/main.bicep:201]:
```suggestion
    database001Name: database001Name
```

File: docs/reference/portal.dataProduct.json [hunk 5]
```diff
@@ -983,6 +1031,8 @@
                 "purviewManagedStorageId": "[if(empty(steps('generalSettings').dataGovernanceSettings.purviewId.id), '', steps('generalSettings').dataGovernanceSettings.purviewApi.properties.managedResources.storageAccount)]",
                 "purviewManagedEventHubId": "[if(empty(steps('generalSettings').dataGovernanceSettings.purviewId.id), '', steps('generalSettings').dataGovernanceSettings.purviewApi.properties.managedResources.eventHubNamespace)]",
                 "enableRoleAssignments": "[steps('generalSettings').generalSettings.enableRoleAssignments]",
+                "enableMonitoring": "[equals(steps('generalSettings').monitoringSettings.enableMonitoring,'yes')]",
+                "dataProductTeamEmail": "[if(empty(steps('generalSettings').monitoringSettings.dataProductTeamEmail),'',steps('generalSettings').monitoringSettings.dataProductTeamEmail)]",
```
Comment (marvinbuss) [inline docs/reference/portal.dataProduct.json:1035]:
```suggestion
                "enableMonitoring": "[equals(steps('generalSettings').monitoringSettings.enableMonitoring, 'yes')]",
                "dataProductTeamEmail": "[if(empty(steps('generalSettings').monitoringSettings.dataProductTeamEmail), '', steps('generalSettings').monitoringSettings.dataProductTeamEmail)]",
```

File: infra/modules/services/synapse.bicep [hunk 6]
```diff
@@ -257,3 +257,8 @@ resource synapsePrivateEndpointDevARecord 'Microsoft.Network/privateEndpoints/pr
 
 // Outputs
 output synapseId string = synapse.id
+output synapseSqlPool001 object = synapseSqlPool001
```
Comment (marvinbuss) [inline infra/modules/services/synapse.bicep:260]: Is this anywhere used?
If not, can we remove it please?

File: infra/modules/services/sql.bicep [hunk 7]
```diff
@@ -71,7 +72,7 @@ resource sqlserverAdministrators 'Microsoft.Sql/servers/administrators@2020-11-0
 
 resource sqlserverDatabase001 'Microsoft.Sql/servers/databases@2020-11-01-preview' = {
   parent: sqlserver
-  name: 'Database001'
+  name: Database001Name
```
Comment (marvinbuss) [inline infra/modules/services/sql.bicep:75]:
```suggestion
  name: database001Name
```

File: infra/modules/services/sql.bicep [hunk 8]
```diff
@@ -15,6 +15,7 @@ param administratorPassword string
 param sqlserverAdminGroupName string = ''
 param sqlserverAdminGroupObjectID string = ''
 param privateDnsZoneIdSqlServer string = ''
+param Database001Name string
```
Comment (marvinbuss) [inline infra/modules/services/sql.bicep:18]:
```suggestion
param database001Name string
```

File: infra/modules/services/diagnosticsettings.bicep [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+    ]
+    metrics: [
+      {
+        category: 'Requests'
+        enabled: true
+      }
+    ]
+  }
+}
+//Outputs
```
Comment (marvinbuss) [inline infra/modules/services/diagnosticsettings.bicep:318]:
```suggestion
}

//Outputs
```

File: infra/modules/services/diagnosticsettings.bicep [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+      {
+        category: 'BigDataPoolAppsEnded'
+        enabled: true
+      }
+    ]
+  }
+}]
+
+// Diagnostic settings for Azure SQL Server.
+resource diagnosticSetting005 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = [for i in range(0, sqlServerDatabasesCount):if (sqlFlavour == 'mysql') {
```
Comment (marvinbuss) [inline infra/modules/services/diagnosticsettings.bicep:187]:
```suggestion
resource diagnosticSetting005 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = [for i in range(0, sqlServerDatabasesCount): if (sqlFlavour == 'mysql') {
```

File: infra/modules/services/diagnosticsettings.bicep [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+      {
+        category: 'Waits'
+        enabled: true
+      }
+    ]
+  }
+}]
+
+// Diagnostic settings for Azure Syanpse Apache Spark Pools. 
+resource diagnosticSetting004 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = [for i in range(0, synapseSparkPoolCount):if (processingService == 'synapse') {
```
Comment (marvinbuss) [inline infra/modules/services/diagnosticsettings.bicep:172]:
```suggestion
resource diagnosticSetting004 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = [for i in range(0, synapseSparkPoolCount): if (processingService == 'synapse') {
```

File: infra/modules/services/diagnosticsettings.bicep [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+      {
+        category: 'IntegrationTriggerRuns'
+        enabled: true
+      }
+    ]
+  }
+}
+
+// Diagnostic settings for Azure Synapse SQL Pools.
+resource diagnosticSetting003 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = [for i in range(0, synapseSqlPoolsCount):if (processingService == 'synapse') {
```
Comment (marvinbuss) [inline infra/modules/services/diagnosticsettings.bicep:141]:
```suggestion
resource diagnosticSetting003 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = [for i in range(0, synapseSqlPoolsCount): if (processingService == 'synapse') {
```

File: infra/modules/services/dashboard.bicep [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+param datafactoryScope string
+param datafactoryName string
+param processingService string
+param synapseScope string
+param synapse001Name string
+
+// Variables
+
+// Resources
+resource dashboard 'Microsoft.Portal/dashboards@2020-09-01-preview' = if (processingService == 'dataFactory') {
```
Comment (marvinbuss) [inline infra/modules/services/dashboard.bicep:20]:
```suggestion
resource dashboardDataFactory 'Microsoft.Portal/dashboards@2020-09-01-preview' = if (processingService == 'dataFactory') {
```

File: infra/modules/services/dashboard.bicep [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+  name: dashboardName
```
Comment (marvinbuss) [inline infra/modules/services/dashboard.bicep:21]:
```suggestion
  name: '${dashboardName}-dataFactory'
```

File: infra/modules/services/cosmosdb.bicep [hunk 15]
```diff
@@ -55,7 +55,7 @@ resource cosmosdb 'Microsoft.DocumentDB/databaseAccounts@2021-03-15' = {
       {
         locationName: location
         failoverPriority: 0
-        isZoneRedundant: true
+        isZoneRedundant: false
```
Comment (marvinbuss) [inline infra/modules/services/cosmosdb.bicep:58]: Why was this disabled?

File: infra/modules/services/alerts.bicep [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+    emailReceivers: [
+      {
+        emailAddress: dataProductTeamEmail
+        name: 'emailaction'
+        useCommonAlertSchema: true
+      }
+    ]
+    enabled: true
+  }
+}
```
Comment (marvinbuss) [inline infra/modules/services/alerts.bicep:39]: What happens if no email was provided here?

File: infra/main.bicep [hunk 17]
```diff
@@ -236,4 +250,77 @@ module postgresql001 'modules/services/postgresql.bicep' = if (sqlFlavour == 'po
...
+    cosmosRequestLimitedAlertName: cosmosRequestLimitedAlertName
+    synapseScope: synapse001.outputs.synapseId
+    cosmosDBScope: cosmosdb001.outputs.cosmosId
+    enableCosmos: enableCosmos
+    tags: tagsJoined
+  }
+}
+
+module dashboard './modules/services/dashboard.bicep' = if (enableMonitoring) {
+  name: 'dashboard'
```
Comment (marvinbuss) [inline infra/main.bicep:312]:
```suggestion
module dashboards './modules/services/dashboard.bicep' = if (enableMonitoring) {
  name: 'dashboards'
```

**Review by marvinbuss — CHANGES_REQUESTED** [review]
Please see comments below.

File: docs/reference/portal.dataProduct.json (hunk 1 above)
Comment (sudivate) [reply docs/reference/portal.dataProduct.json]: Fixed

File: infra/modules/services/alerts.bicep (hunk 16 above)
Comment (sudivate) [reply infra/modules/services/alerts.bicep:39]: Alert module has a check for email

File: infra/main.bicep [hunk 18]
```diff
@@ -97,10 +103,17 @@ var keyVault001Name = '${name}-vault001'
 var synapse001Name = '${name}-synapse001'
 var datafactory001Name = '${name}-datafactory001'
 var cosmosdb001Name = '${name}-cosmos001'
+var database001Name = 'Database001'
 var sql001Name = '${name}-sqlserver001'
 var mysql001Name = '${name}-mysql001'
 var mariadb001Name = '${name}-mariadb001'
 var potsgresql001Name = '${name}-postgresql001'
+var logAnalytics001Name = '${name}-loganalytics'
```
Comment (marvinbuss) [inline infra/main.bicep:111]:
```suggestion
var logAnalytics001Name = '${name}-loganalytics001'
```

File: infra/main.bicep [hunk 19]
```diff
@@ -97,10 +103,17 @@ var keyVault001Name = '${name}-vault001'
...
+var dataFactoryEmailActionGroup = '${datafactory001Name}-${name}-emailactiongroup'
+var adfPipelineFailedAlertName = '${datafactory001Name}-${name}-failedalert'
```
Comment (marvinbuss) [inline infra/main.bicep:113]:
```suggestion
var dataFactoryEmailActionGroup = '${datafactory001Name}-emailactiongroup'
var adfPipelineFailedAlertName = '${datafactory001Name}-failedalert'
```

File: infra/modules/services/loganalytics.bicep [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+// Licensed under the MIT license.
+
+// This template is used to create a Log Analytics workspace.
+targetScope = 'resourceGroup'
+
+// Parameters
+param location string
+param tags object
+param processingService string
+param logAnalytics001Name string
```
Comment (marvinbuss) [inline infra/modules/services/loganalytics.bicep:11]:
```suggestion
param logAnalyticsName string
```

File: infra/modules/services/loganalytics.bicep [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+param sqlFlavour string
+
+// Variables
+var dataFactoryAnalyticsName = 'AzureDataFactoryAnalytics(${logAnalytics001Name})'
+var sqlAnalyticsName = 'AzureSQLAnalytics(${logAnalytics001Name})'
```
Comment (marvinbuss) [inline infra/modules/services/loganalytics.bicep:16]:
```suggestion
var dataFactoryAnalyticsName = 'AzureDataFactoryAnalytics(${logAnalyticsName})'
var sqlAnalyticsName = 'AzureSQLAnalytics(${logAnalyticsName})'
```

File: infra/main.bicep [hunk 22]
```diff
@@ -236,4 +250,77 @@ module postgresql001 'modules/services/postgresql.bicep' = if (sqlFlavour == 'po
   }
 }
 
+module logAnalytics001 'modules/services/loganalytics.bicep' = if (enableMonitoring) {
+  name: 'logAnalytics001'
+  scope: resourceGroup()
+  params: {
+    location: location
+    tags: tagsJoined
+    logAnalytics001Name: logAnalytics001Name
```
Comment (marvinbuss) [inline infra/main.bicep:259]:
```suggestion
    logAnalyticsName: logAnalytics001Name
```

**Review by marvinbuss — APPROVED** [review]
LGTM!

**Comment by marvinbuss** [comment]
@sudivate Can you please check whether one of the requests has been considered?

**Review by marvinbuss — APPROVED** [review]
LGTM!

**Pull request merged**